src/common/bots/Bot.cpp

```diff
@@ -0,0 +1,110 @@
+#include "Bot.hpp"
+
+Bot::Bot() {};
+
+std::pair<InputDef::ReliableInput, InputDef::UnreliableInput>
+    Bot::vectorToNearestMoveTransform(sf::Vector2f v) {
+    InputDef::ReliableInput reliable_input;
+    reliable_input.setAll(false);
+
+    v = VectorUtil::normalize(v);
+    std::array<sf::Vector2f, 8> directions = {
+        sf::Vector2f(0.f,-1.f),  //up
+        sf::Vector2f(0.f,1.f),   //down
+        sf::Vector2f(-1.f,0.f),  //left
+        sf::Vector2f(1.f,0.f),   //right
+        sf::Vector2f(-1.f,-1.f), //left-up
+        sf::Vector2f(1.f,1.f),   //right-down
+        sf::Vector2f(-1.f,1.f),  //left-down
+        sf::Vector2f(1.f,-1.f),  //right-up
+    };
+    sf::Vector2f closestAllowedDirection = *std::max(directions.begin(), directions.end(),
+        [v](sf::Vector2f* dir_a, sf::Vector2f* dir_b) {
+            return VectorUtil::dot(v, *dir_a) > VectorUtil::dot(v, *dir_b);
+        }
+    );
+
+    bool go_left = closestAllowedDirection.x == -1;
+    bool go_right = closestAllowedDirection.x == 1;
+    bool go_up = closestAllowedDirection.y == -1;
+    bool go_down = closestAllowedDirection.y == 1;
+    InputDef::UnreliableInput unreliable_input = {
+        .toggle_up = go_up,
+        .toggle_down = go_down,
+        .toggle_right = go_right,
+        .toggle_left = go_left,
+        .toggle_stop = false,
+    };
+    return std::pair<InputDef::ReliableInput, InputDef::UnreliableInput>(reliable_input,
+                                                                         unreliable_input);
+}
+
+/* This strategy will move towards the closest mine which has a resource the bot needs to win.
+ * Once it has enough of a resource, it will move to the next closest mine with a resource it needs. */
+std::pair<InputDef::ReliableInput, InputDef::UnreliableInput>
+Bot::getNearestGreedyMove(std::vector<std::shared_ptr<Mine>> mines,
+                std::array<uint32_t, RESOURCE_TYPE_COUNT> current_resources,
+                sf::Vector2f bot_center,
+                std::unordered_map<uint32_t, uint32_t> mine_id_to_resource_count
+                ) {
+
+
+    // how expensive is this? We really only need to do it when we change direction, since
+    // if we know we've been moving towards a mine we can assume we're continuing to move towards it.
+    // in which case we can't possibly have gotten closer to a different mine.
```

Comment:
yeah keeping state that indicates whether you already have a goal (mine to move towards) will save you a lot of computation here. The goal can reset to None when you satisfy it or something

Reply:
(n log n) * calculating vector distance 
https://stackoverflow.com/questions/4484900/what-is-the-time-complexity-of-stdsort-in-the-c-standard-library

Reply:
yeah Ive been thinking about his, probably won't optimize given N is just 4 here and it's not a noticeable issue.

Reply:
yeah whatever